cub/test/catch2_test_device_histogram.cu

```diff
@@ -0,0 +1,748 @@
+  const array<LevelT, ActiveChannels>& lower_level,
+  const array<LevelT, ActiveChannels>& upper_level,
+  OffsetT num_row_pixels,
+  OffsetT num_rows,
+  OffsetT row_stride_bytes,
+  std::false_type = {})
+{
+  LevelT fp_scales[ActiveChannels]; // only used when LevelT is floating point
+  (void) fp_scales; // TODO(bgruber): use [[maybe_unsued]] in C++17
+
+  for (int c = 0; c < ActiveChannels; ++c)
+  {
+    _CCCL_IF_CONSTEXPR (!std::is_integral<LevelT>::value)
+    {
+      fp_scales[c] =
+        LevelT{1} / static_cast<LevelT>((upper_level[c] - lower_level[c]) / static_cast<LevelT>(num_levels[c] - 1));
+    }
+  }
+
+  auto sample_to_bin_index = [&](int channel, SampleT sample) {
+    const auto n   = num_levels[channel];
+    const auto max = upper_level[channel];
+    const auto min = lower_level[channel];
+    if (sample < min || sample >= max)
+    {
+      return n; // Sample out of range
+    }
+    _CCCL_IF_CONSTEXPR (std::is_integral<LevelT>::value)
+    {
+      // Accurate bin computation following the arithmetic we guarantee in the HistoEven docs
+      return static_cast<int>(
+        static_cast<uint64_t>(sample - min) * static_cast<uint64_t>(n - 1) / static_cast<uint64_t>(max - min));
+    }
+    else
+    {
+      return static_cast<int>((static_cast<float>(sample) - min) * fp_scales[channel]);
+    }
+    _LIBCUDACXX_UNREACHABLE();
+  };
+  auto h_histogram = compute_reference_result<Channels, ActiveChannels, CounterT>(
+    h_samples, sample_to_bin_index, num_levels, num_row_pixels, num_rows, row_stride_bytes);
+
+  // Allocate and initialize device data
+  c2h::device_vector<CounterT> d_histogram[ActiveChannels];
+  for (int channel = 0; channel < ActiveChannels; ++channel)
+  {
+    d_histogram[channel].resize(num_levels[channel] - 1);
+  }
+
+  // TODO(bgruber): replace by launch wrapper, which handles allocation and calling CUB APIs
+
+  size_t temp_storage_bytes = 0;
+  call_even<ActiveChannels, Channels>(
+    nullptr,
+    temp_storage_bytes,
+    d_samples,
+    d_histogram,
+    num_levels,
+    lower_level,
+    upper_level,
+    num_row_pixels,
+    num_rows,
+    row_stride_bytes);
+  constexpr char canary_token = 8;
+  const auto canary_zone      = c2h::host_vector<char>(256, canary_token);
+  auto d_temp_storage         = c2h::device_vector<char>(temp_storage_bytes + canary_zone.size() * 2, canary_token);
+
+  // Run warmup/correctness iteration
+  call_even<ActiveChannels, Channels>(
+    thrust::raw_pointer_cast(d_temp_storage.data()) + canary_zone.size(),
+    temp_storage_bytes,
+    d_samples,
+    d_histogram,
+    num_levels,
+    lower_level,
+    upper_level,
+    num_row_pixels,
+    num_rows,
+    row_stride_bytes);
+
+  // Check canary zones and channel histograms
+  CHECK(c2h::host_vector<char>(d_temp_storage.begin(), d_temp_storage.begin() + canary_zone.size()) == canary_zone);
+  CHECK(c2h::host_vector<char>(d_temp_storage.end() - canary_zone.size(), d_temp_storage.end()) == canary_zone);
```

Comment:
@gevtushenko One of the biggest obstacles to using launch wrappers is now the testing of out-of-bounds writes to the temporary storage. The old tests allocated 256 bytes left and right of the temporary storage, filled them with a pattern, and verified the pattern was unchanged by the kernel. Can we drop this kind of verification? I feel such memory errors should rather be tested for by tools like compute-sanititzer, and not manually implemented.

Reply:
Alternatively, we could add this kind of verification to all tests, by moving the checks into the launch wrapper.

Reply:
@bernhardmgruber I believe histogram is no different in handling temporary storage from other algorithms. I'd suggest to drop this check for now and file an issue on investigating ideas about memory check CI integration vs specialized API for canary bits in launch wrapper.

Reply:
The checks have been dropped. Here is the issue to follow-up with more general approaches: #1775